### DISCUSSION TITLE
Why is CS9035 raised for records with primary ctor and required properties?

### DISCUSSION THREAD

**Discussion by julealgon** [dotnet/csharplang#6780] [Q&A]
I was making a few tests here with the new `required` modifier and noticed what I believe to be an incorrect emission of the CS9035 warning.

```csharp
_ = new Y();
_ = new Y("value");
_ = new Y { Value = "value" };

readonly record struct Y(string Value)
{
    public required string Value { get; init; } = Value;
}
```

| Severity | Code | Description | Line |
| -- | -- | -- | -- |
| Error | CS9035 | Required member 'Y.Value' must be set in the object initializer or attribute constructor.	| 1 |
| Error | CS9035 | Required member 'Y.Value' must be set in the object initializer or attribute constructor. | 2 |

The first warning works as I expect as I'm not providing `Value` on line1.
The second warning however looks wrong to me. I'm passing `Value` and using the same name to initialize my property 

The docs don't seem to talk explicitly about `required`, but [they mention this](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#positional-syntax-for-property-definition:~:text=If%20the%20generated,initialize%20the%20property.):

> If the generated auto-implemented property definition isn't what you want, you can define your own property of the same name. For example, you may want to change accessibility or mutability, or provide an implementation for either the get or set accessor. If you declare the property in your source, you must initialize it from the positional parameter of the record. If your property is an auto-implemented property, you must initialize the property.

It would be even better if this was allowed though:
```csharp
readonly record struct Y(required string Value)
```
However that just fails spectacularly with a completely unrelated compilation error:
![image](https://user-images.githubusercontent.com/461579/205746045-83e2bc13-92cc-4c6d-b9be-e78fd5b342e0.png)

I do understand I can just opt-out of the primary constructor, but that also strips away the automatically generated tuple deconstructor, which I do want.

To get the behavior I want, I'm forced to drop the primary constructor and add my own deconstruct logic.

**Comment by 333fred** [comment]
This behaves exactly as intended, which is why primary constructor members are not defined as `required` by default.

**Reply by julealgon** [reply]
Wait. Are you saying it is intended to be impossible to set record properties as `required` when using primary constructors then?

That feels very arbitrary to me and I'd like to see that change it possible. It is also not orthogonal in regards to how I can "override" other aspects of the generated property by redefining it.

**Reply by 333fred** [reply]
> Are you saying it is intended to be impossible to set record properties as required when using primary constructors then?

No, you _are_ allowed to add `required`, as you did above. I am saying that if you do make such properties `required`, that you will then be required to set them in an object initializer, just like any other constructor. This then results in exactly the behavior you are experiencing, and that is intended.

**Reply by HaloFour** [reply]
What makes this wonky is that if the property is `required` then it must be set in the initializer, even if the value is already set in the constructor.  From a consumption point of view, the compiler has no idea that the `Value` parameter in the constructor and the `Value` property are related.  Such a constructor would need to be annotated with the `SetsRequiredMembersAttribute`, which I don't think is possible with primary constructors?

Ultimately you're probably better off writing your own constructor and deconstruction logic:

```cs
using System.Diagnostics.CodeAnalysis;

_ = new Y(); // error
_ = new Y("value"); // no error
_ = new Y { Value = "value" }; // no error


readonly record struct Y
{
    public required string Value { get; init; }
    
    [SetsRequiredMembers]
    public Y(string value) {
        this.Value = value;
    }
    
    public void Deconstruct(out string value) {
        value = this.Value;
    }
}
```

**Reply by julealgon** [reply]
> the compiler has no idea that the `Value` parameter in the constructor and the `Value` property are related.

What about the fact that I'm using the exact same name as in the ctor to set the property value? Wouldn't that be able to be used to check initialization happened?

**Reply by HaloFour** [reply]
> What about the fact that I'm using the exact same name as in the ctor to set the property value? Wouldn't that be able to be used to check initialization happened?

What happens in the constructor is opaque to a consumer of that type.

**Reply by julealgon** [reply]
> What happens in the constructor is opaque to a consumer of that type.

I can see how that applies to normal constructors where you have arbitrary logic inside, but for primary constructors though? Wouldn't the compiler be able to tell you are initializing the properties in that scenario?

Also, how can the compiler tell that you have initialized a `readonly` property today? What's different in checking whether a `readonly` member was assigned to vs a `required` one?

**Reply by HaloFour** [reply]
> I can see how that applies to normal constructors where you have arbitrary logic inside, but for primary constructors though?

To an outside assembly the fact that the constructor was a primary constructor is entirely lost.  Most of the metadata that would imply that the type is a record is too.  There would need to be additional metadata that specifies that the constructor satisfies initialization of those required properties, which is exactly what `SetsRequiredMembersAttribute` does.

> Also, how can the compiler tell that you have initialized a `readonly` property today?  What's different in checking whether a `readonly` member was assigned to vs a `required` one?

You mean a property with an `init` accessor?  All the compiler needs to do there is enforce that the only time that you can write to that member is during initialization.  The compiler has no idea whether or not that member had already been written to within a constructor.

**Reply by TahirAhmadov** [reply]
~~I guess the question is, why aren't all primary ctors annotated with `[SetsRequiredMembers]`?~~

NVM, I figured it out.
